multi/imxrt-multi/spi.c

```diff
@@ -0,0 +1,441 @@
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <sys/interrupt.h>
+#include <sys/threads.h>
+#include <sys/pwman.h>
+#include <sys/msg.h>
+#include <unistd.h>
+#include <errno.h>
+
+
+#include "imxrt-multi.h"
+#include "common.h"
+#include "config.h"
+#include "spi.h"
+
+#define SPI1_POS 0
+#define SPI2_POS (SPI1_POS + SPI1)
+#define SPI3_POS (SPI2_POS + SPI2)
+#define SPI4_POS (SPI3_POS + SPI3)
+
+#define SPI_CNT (SPI1 + SPI2 + SPI3 + SPI4)
+
+#define MAX_FRAME_SZ 0x1000
+#define WORD_SIZE sizeof(uint32_t)
+#define MAX_FIFOSZ_BYTES 16 * WORD_SIZE
+
+#define DESERIALIZE_WORD(buff) \
+	buff[3] | (buff[2] << 8) | (buff[1] << 16) | (buff[0] << 24)
+
+#define SERIALIZE_WORD(buff, word, start)		\
+	for (int i = start - 1; i >= 0; --i) 		\
+		*buff++ = (word >> (i * 8)) & 0xff;
+
+
+enum { spi_verid = 0, spi_param, spi_cr = 0x4, spi_sr, spi_ier, spi_der, spi_cfgr0, spi_cfgr1, spi_dmr0 = 0xc,
+	   spi_dmr1, spi_ccr = 0x10, spi_fcr = 0x16, spi_fsr, spi_tcr, spi_tdr, spi_rsr = 0x1c, spi_rdr };
+
+
+struct {
+	handle_t cond;
+	handle_t mutex;
+	handle_t inth;
+	handle_t irqLock;
+
+	int irq;
+	volatile int ready;
+	volatile uint32_t *base;
+
+	uint32_t tcr;
+} spi_common[SPI_CNT];
+
+
+static const int spiConfig[] = { SPI1, SPI2, SPI3, SPI4 };
+
+
+static const int spiPos[] = { SPI1_POS, SPI2_POS, SPI3_POS, SPI4_POS };
+
+
+static int spi_irqHandler(unsigned int n, void *arg)
+{
+	int spi = (int)arg;
+	*(spi_common[spi].base + spi_ier) &= ~(1 << 0);
+	spi_common[spi].ready = 1;
+
+	return 0;
+}
+
+
+static void spi_initTransmition(int spi)
+{
+	mutexLock(spi_common[spi].irqLock);
+	spi_common[spi].ready = 0;
+
+	*(spi_common[0].base + spi_ier) |= 1;
+
+	while (!spi_common[spi].ready)
+		condWait(spi_common[spi].cond, spi_common[spi].irqLock, 0);
+
+	mutexUnlock(spi_common[spi].irqLock);
+}
+
+
+static void spi_txBytes(int spi, const uint8_t *txBuff, int bytesNumber)
+{
+	uint32_t word;
+
+	while (bytesNumber / WORD_SIZE) {
+		*(spi_common[spi].base + spi_tdr) = DESERIALIZE_WORD(txBuff);
+		txBuff += WORD_SIZE;
+		bytesNumber -= WORD_SIZE;
+	}
+
+	if (bytesNumber) {
+		word = 0;
+		while (bytesNumber--)
+			word = (word << 8) | *txBuff++;
+		*(spi_common[spi].base + spi_tdr) = word;
+	}
+}
+
+
+static void spi_rxBytes(int spi, uint8_t *rxBuff, int bytesNumber)
+{
+	uint32_t word;
+
+	/* Get data from RX Fifo */
+	while (bytesNumber / WORD_SIZE) {
+		word = *(spi_common[spi].base + spi_rdr);
+		SERIALIZE_WORD(rxBuff, word, WORD_SIZE);
+		bytesNumber -= WORD_SIZE;
+	}
+
+	if (bytesNumber) {
+		word = *(spi_common[spi].base + spi_rdr);
+		SERIALIZE_WORD(rxBuff, word, bytesNumber);
+	}
+}
+
+
+static int spi_transmitData(int spi, const uint8_t *txBuff, uint8_t *rxBuff, int len)
+{
+	int size = len;
+	int rxTotalBytes = 0;
+	int txFifoBytes, rxFifoBytes;
+	volatile uint32_t txWordsCnt, rxWordsCnt;
+
+	if (!spiConfig[spi])
+		return -EINVAL;
+
+	if ((len * 8) > MAX_FRAME_SZ)
+		return -EINVAL;
+
```

Comment:
I see no mutex in this function. Please have in mind that more than one peer can want access to the same spi at the same time